libraries/standard/mqtt/src/mqtt.c

```diff
@@ -99,6 +101,427 @@ static int32_t sendPacket( MQTTContext_t * pContext,
+                             "Incoming packet will be dumped",
+                             incomingPacket.remainingLength
+                             pContext->networkBuffer );
+    }
+    else
+    {
+        bytesToReceive = incomingPacket.remainingLength;
+    }
+
+    bytesReceived = pContext->transportInterface.recv( pContext->transportInterface.networkContext,
+                                                       pContext->networkBuffer.pBuffer,
+                                                       bytesToReceive );
+    totalBytesReceived = bytesReceived;
+
+    /* Partial receive. */
+    if( bytesReceived < ( int32_t ) bytesToReceive )
+    {
+        IotLogWarnWithArgs( "Packet partially received: Bytes received=%d,"
+                            "Bytes expected=%u",
+                            bytesReceived,
+                            bytesToReceive );
+        /* Handle partial receive. */
+        bytesReceived = pContext->transportInterface.recv( pContext->transportInterface.networkContext,
+                                                           ( pContext->networkBuffer.pBuffer + totalBytesReceived ),
+                                                           ( bytesToReceive - totalBytesReceived ) );
+        totalBytesReceived += bytesReceived;
+        if( totalBytesReceived != ( int32_t ) bytesToReceive )
+        {
+            IotLogErrorWithArgs( "Handling partial receive failed. Bytes "
+                                 "received=%d, Bytes expected=%u",
+                                 totalBytesReceived,
+                                 bytesToReceive );
+            status = MQTTRecvFailed;
+        }
+    }
+    /* Too many bytes received. */
+    else if( bytesReceived > ( int32_t ) bytesToReceive )
+    {
+        IotLogErrorWithArgs( "Too many bytes received. Bytes received=%d, ",
+                             "Bytes expected=%u",
+                             bytesReceived,
+                             bytesToReceive );
+        status = MQTTRecvFailed;
+    }
+    else
+    {
+        /* Receive successful, bytesReceived == bytesToReceive. */
+        IotLogInfoWithArgs( "Packet received. Received bytes=%d",
+                            bytesReceived );
+    }
+
+    /* Check if packet exceeds buffer. */
+    if( ( status == MQTTSuccess ) && ( incomingPacket.remainingLength > bytesToReceive ) )
+    {
+        /* Packet exceeds buffer, dump it. */
+        status = dumpPacket( pContext, incomingPacket.remainingLength, totalBytesReceived );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+/**
+ * @brief Send acks for received QoS 1/2 publishes.
+ *
+ * @param[in] pContext MQTT Connection context.
+ * @param[in] packetId, packet ID of original PUBLISH.
+ * @param[in,out] pPublishState Current/updated publish state in record.
+ *
+ * @return MQTTSuccess or MQTTIllegalState.
+ */
+static MQTTStatus_t sendPublishAcks( MQTTContext_t * const pContext,
```

Comment:
`MQTTContext_t * const` represents top-level constness which does carry any value (as the pointer-value is copied into the parameter). I suggest removing the top-level constness to just use `MQTTContext_t * `

Reply:
The `const` is unnecessary, but I believe it was added for MISRA compliance.

Reply:
Can you share the MISRA violation? I have usually seen MISRA violations that suggest constness only for the underlying type when the pointed to memory is not modified in the function (i.e. in this case as `const MQTTContext_t *` type)